phoebe/solverbackends/solverbackends.py

```diff
@@ -2247,6 +2261,240 @@ def valid_options(self):
+                    {'qualifier': 'fitted_units', 'value': [u if isinstance(u, str) else u.to_string() for u in fitted_units]},
+                    {'qualifier': 'adopt_parameters', 'value': params_twigs, 'choices': params_twigs}]
+
+            if kwargs.get('expose_lnprobabilities', False):
+                custom_lnprobability_callable = kwargs.get('custom_lnprobability_callable')
+
+                if custom_lnprobability_callable is None:
+                    initial_lnprobability = b_solver.calculate_lnlikelihood(model='baseline', consider_gaussian_process=False)
+                else:
+                    initial_lnprobability = custom_lnprobability_callable(b_solver, model='baseline')
+                if _use_progressbar:
+                    _dc_pbar.update(1)
+
+                args = (b_solver, params_uniqueids, compute, [], None, kwargs.get('solution', None), compute_kwargs, custom_lnprobability_callable)
+                fitted_lnprobability = _lnprobability(p0+corrections, *args)
+
+                return_ += [{'qualifier': 'initial_lnprobability', 'value': initial_lnprobability},
+                             {'qualifier': 'fitted_lnprobability', 'value': fitted_lnprobability}]
+
+            return [return_]
+
+        # set _within solver to prevent run_compute progressbars
+        b._within_solver = True
+
+        if kwargs.get('progressbar', False):
+            nforward_models = 1  # baseline model
+            nanalytic_params = len(fitted_params_ps.filter(qualifier=['pblum'], **_skip_filter_checks).twigs)
+            if kwargs.get('deriv_method') == 'symmetric':
+                nforward_models += 2*(len(params_uniqueids)-nanalytic_params)
+            else:
+                nforward_models += len(params_uniqueids)-nanalytic_params
+            if kwargs.get('expose_lnprobabilities'):
+                nforward_models += 1
+            _dc_pbar = _tqdm(total=nforward_models)
+            _use_progressbar = True
+        else:
+            _use_progressbar = False
+
+        datasets_enabled = b.filter(qualifier='enabled', value=True, compute=compute, context='compute', **_skip_filter_checks).datasets
+        # limit to LCs/RVs
+        datasets_enabled = b.filter(dataset=datasets_enabled, kind=['lc', 'rv'], context='dataset', **_skip_filter_checks).datasets
+
+        # need to ensure we access the obs and sigma arrays in the same order
+        obs_params = b.filter(qualifier=['fluxes', 'rvs'], dataset=datasets_enabled, context='dataset', **_skip_filter_checks).to_list()
+        obs_data = np.concatenate([p.get_value() for p in obs_params])
+        obs_sigmas = np.concatenate([b.get_parameter(qualifier=['sigmas'], dataset=p.dataset, component=p.component, context='dataset', **_skip_filter_checks).get_value() for p in obs_params])
+
+        if len(obs_sigmas) != len(obs_data):
+            raise ValueError("sigmas must be provided for all enabled lc and rv datasets")
```

Comment:
It's quite incredible to see how much boilerplate code is required up to this point. 8-|